File: starters/php/src/Session/SessionMachine.php
<?php

declare(strict_types=1);

namespace FProto\Session;

final class SessionMachine
{
    /** @var array<string, TransitionResult> */
    private static array $transitions = [];

    private static function init(): void
    {
        if (!empty(self::$transitions)) {
            return;
        }
        $t = static fn(SessionStatus $s, SessionEvent $e): string => $s->value . ':' . $e->value;

        self::$transitions[$t(SessionStatus::Connecting, SessionEvent::HandshakeOK)] =
            new TransitionResult(SessionStatus::Active, false);
        self::$transitions[$t(SessionStatus::Connecting, SessionEvent::HandshakeFail)] =
            new TransitionResult(SessionStatus::Expired, true);
        self::$transitions[$t(SessionStatus::Active, SessionEvent::PacketReceived)] =
            new TransitionResult(SessionStatus::Active, false);
        self::$transitions[$t(SessionStatus::Active, SessionEvent::IdleTimeout)] =
            new TransitionResult(SessionStatus::Sleeping, false);
        self::$transitions[$t(SessionStatus::Active, SessionEvent::ExplicitLogout)] =
            new TransitionResult(SessionStatus::Expired, false);
        self::$transitions[$t(SessionStatus::Sleeping, SessionEvent::PacketReceived)] =
            new TransitionResult(SessionStatus::Active, false);
        self::$transitions[$t(SessionStatus::Sleeping, SessionEvent::ExpiryTimeout)] =
            new TransitionResult(SessionStatus::Expired, false);
    }

    public static function transition(SessionStatus $current, SessionEvent $event): TransitionResult
    {
        self::init();
        $key = $current->value . ':' . $event->value;
        if (!isset(self::$transitions[$key])) {
            throw new InvalidTransitionException($current, $event);
        }
        return self::$transitions[$key];
    }
}

File: starters/java/src/main/java/io/fproto/core/SessionMachine.java
package io.fproto.core;

import java.util.Map;

public final class SessionMachine {

    private SessionMachine() {}

    private record Key(SessionStatus status, SessionEvent event) {}

    private static final Map<Key, TransitionResult> TRANSITIONS = Map.of(
        new Key(SessionStatus.CONNECTING, SessionEvent.HANDSHAKE_OK),
            new TransitionResult(SessionStatus.ACTIVE, false),
        new Key(SessionStatus.CONNECTING, SessionEvent.HANDSHAKE_FAIL),
            new TransitionResult(SessionStatus.EXPIRED, true),
        new Key(SessionStatus.ACTIVE, SessionEvent.PACKET_RECEIVED),
            new TransitionResult(SessionStatus.ACTIVE, false),
        new Key(SessionStatus.ACTIVE, SessionEvent.IDLE_TIMEOUT),
            new TransitionResult(SessionStatus.SLEEPING, false),
        new Key(SessionStatus.ACTIVE, SessionEvent.EXPLICIT_LOGOUT),
            new TransitionResult(SessionStatus.EXPIRED, false),
        new Key(SessionStatus.SLEEPING, SessionEvent.PACKET_RECEIVED),
            new TransitionResult(SessionStatus.ACTIVE, false),
        new Key(SessionStatus.SLEEPING, SessionEvent.EXPIRY_TIMEOUT),
            new TransitionResult(SessionStatus.EXPIRED, false)
    );

    public static TransitionResult transition(SessionStatus current, SessionEvent event)
            throws InvalidTransitionException {
        var key = new Key(current, event);
        var result = TRANSITIONS.get(key);
        if (result == null) {
            throw new InvalidTransitionException(current, event);
        }
        return result;
    }
}
